internal/provider/resource_gitlab_branch.go

```diff
@@ -0,0 +1,254 @@
+	branch, resp, err := client.Branches.CreateBranch(project, branchOptions, gitlab.WithContext(ctx))
+	if err != nil {
+		log.Printf("[DEBUG] failed to create gitlab branch %v response %v", branch, resp)
+		return diag.FromErr(err)
+	}
+	d.SetId(buildTwoPartID(&project, &name))
+	return resourceGitlabBranchRead(ctx, d, meta)
+}
+
+func resourceGitlabBranchRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	project, name, err := parseTwoPartID(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	log.Printf("[DEBUG] read gitlab branch %s", name)
+	branch, resp, err := client.Branches.GetBranch(project, name, gitlab.WithContext(ctx))
+	if err != nil {
+		if is404(err) {
+			log.Printf("[DEBUG] recieved 404 for gitlab branch %s, removing from state", name)
+			d.SetId("")
+			return diag.FromErr(err)
+		}
+		log.Printf("[DEBUG] failed to read gitlab branch %s response %v", name, resp)
+		return diag.FromErr(err)
+	}
+	ref := d.Get("ref").(string)
+	if ref == "" {
+		c := &gitlab.GetCommitRefsOptions{
+			Type: gitlab.String("branch"),
+		}
+		commitRefs, _, err := client.Commits.GetCommitRefs(project, branch.Commit.ID, c, gitlab.WithContext(ctx))
+		if err != nil {
+			return diag.FromErr(err)
+		}
+		for _, br := range commitRefs {
+			if br.Name != branch.Name {
+				ref = br.Name
+				break
+			}
+		}
+	}
+	d.Set("ref", ref)
+	d.SetId(buildTwoPartID(&project, &name))
+	d.Set("name", branch.Name)
+	d.Set("project", project)
+	d.Set("web_url", branch.WebURL)
+	d.Set("default", branch.Default)
+	d.Set("can_push", branch.CanPush)
+	d.Set("merged", branch.Merged)
+	d.Set("developer_can_merge", branch.DevelopersCanMerge)
+	d.Set("developer_can_push", branch.DevelopersCanPush)
+	d.Set("protected", branch.Protected)
+	commit := flattenCommit(branch.Commit)
+	if err := d.Set("commit", commit); err != nil {
+		return diag.FromErr(err)
+	}
+	return nil
+}
+
+func resourceGitlabBranchDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	project, name, err := parseTwoPartID(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	log.Printf("[DEBUG] delete gitlab branch %s", name)
+	resp, err := client.Branches.DeleteBranch(project, name, gitlab.WithContext(ctx))
+	if err != nil {
+		log.Printf("[DEBUG] failed to delete gitlab branch %s response %v", name, resp)
+		return diag.FromErr(err)
+	}
+	return nil
+}
+
+func flattenCommit(commit *gitlab.Commit) (values []map[string]interface{}) {
+	if commit == nil {
+		return []map[string]interface{}{}
+	}
+	return []map[string]interface{}{
+		{
+			"id":              commit.ID,
+			"short_id":        commit.ShortID,
+			"title":           commit.Title,
+			"author_name":     commit.AuthorName,
+			"author_email":    commit.AuthorEmail,
+			"authored_date":   commit.AuthoredDate.String(),
+			"committed_date":  commit.CommittedDate.String(),
```

Comment:
These should use the format of the GitLab API.
```suggestion
			"authored_date":   commit.AuthoredDate.Format(time.RFC3339),
			"committed_date":  commit.CommittedDate.Format(time.RFC3339),
```